lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,140 @@
+```
+
+In the `metadata.yaml` of the charm, add the following:
+
+```yaml
+requires:
+  pushgateway:
+    interface: pushgateway
+```
+
+In the source of your charm, first import the interface:
+
+```
+from charms.prometheus_pushgateway_k8s.v0.pushgateway import PrometheusPushgatewayInterface
+````
+
+Just instantiate the object in `__init__` and then use it when desired to send a metric, passing
+its name and value:
+
+```
+self.ppi = PrometheusPushgatewayInterface(self)
+...
+self.ppi.send_metric("test_metric", 3.141592)
+```
+
+The `send_metric` call will just end quietly if the metric was sent succesfully, or will raise
+an exception if something is wrong (that error should be logged or informed to the operator).
+
+
+## Waiting for Pushgateway
+
+When your charm is deployed but the relation is still not added to the Prometheus Pushgateway,
+metrics cannot not be sent.
+
+For robustness you should only send metrics when the interface is ready:
+
+```
+self.ppi = PrometheusPushgatewayInterface(self)
+...
+if self.ppi.is_ready:
+    self.ppi.send_metric("test_metric", 3.141592)
+```
+"""
+
+import json
+import logging
+from typing import Union
+
+import requests
+from ops.charm import CharmBase, HookEvent
+from ops.framework import Object, StoredState
+
+logger = logging.getLogger(__name__)
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "a0065690fe484ef296d5847fcbf1d728"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["requests"]
+
+
+class PrometheusPushgatewayInterface(Object):
+    """Interface for the Prometheus Pushgateway."""
+
+    _stored = StoredState()
+
+    def __init__(self, charm: CharmBase):
+        """Construct the interface for the Prometheus Pushgateway.
+
+        Args:
+            charm: a `CharmBase` object that manages this
+                `MetricsEndpointProvider` object. Typically, this is
+                `self` in the instantiating class.
+        """
+        super().__init__(charm, None)
+        self._stored.set_default(pushgateway_url=None)
+
+        self.framework.observe(
+            charm.on.pushgateway_relation_created, self._on_push_relation_changed
+        )
+        self.framework.observe(
+            charm.on.pushgateway_relation_changed, self._on_push_relation_changed
+        )
+        self.framework.observe(
+            charm.on.pushgateway_relation_broken, self._on_push_relation_removed
+        )
+
+    @property
+    def is_ready(self):
+        """Return if the service is ready to send metrics."""
+        return self._stored.pushgateway_url is not None
+
+    def _on_push_relation_changed(self, event: HookEvent) -> None:
+        """Receive the push endpoint information."""
+        raw = event.relation.data[event.app].get("push-endpoint")
```

Comment:
It's safer to add some guards here against `event.relation.data` raising when fetched. The charm could end up processing this event when the relation is being removed and juju has been a bit too hasty in doing so.
I'd wrap the whole line in a `try/except ModelError` to feel safe.

Reply:
On the new model we just get the relation dynamically so we shouldn't hit this race condition.